### ISSUE TITLE
[Feature request] Alibi for arbitrary positions

### ISSUE THREAD

**Issue by Aceticia** [lucidrains/x-transformers#280]
Hi, I've been using this library for a while and I used it for another application where time is not simply described by `arange` due to irregular sampling. Two possible ways to solve this:

1. Modify the alibi module to take an optional distance matrix provided directly by the user
2. User provide an optional time array for key and/or query and compute the distance matrix within alibi module

If you are tied up with other stuff I can maybe try to help contribute this. But please let me know if this is not part of the scope of this project.

**Comment by lucidrains** [comment]
hey Xujin! thanks for bringing this up

I do have plans for this actually because I am meditating on https://arxiv.org/abs/2305.16843

let's keep this issue open!

**Comment by lucidrains** [comment]
@Aceticia hey Xujin, decided to make it so you can create the alibi bias and pass it in externally for starters

in the latest version, try

```python
import torch
from x_transformers import Decoder
from x_transformers.x_transformers import AlibiPositionalBias

encoder = Decoder(
    dim = 512,
    depth = 2,
    heads = 8,
)

x = torch.randn(2, 4, 512)

alibi = AlibiPositionalBias(heads = 8)

pos = torch.tensor([[0, 1, 2, 4], [1, 3, 5, 7]])

attn_bias = alibi.forward_custom_pos(pos)

embed = encoder(x, attn_bias = attn_bias)
```

let me know if that works for your research

**Comment by lucidrains** [comment]
@Aceticia made it more [tightly integrated](https://github.com/lucidrains/x-transformers/commit/7e15c09b8567261799e07ee7b53912ddc6f11f2d#diff-10b0fcd6b3df77bf858789e5e9fb0b0662139dc0b1d66430223123062a503f8cR380)

think this feature should be good to go

**Comment by Aceticia** [comment]
This looks great. Thank you so much. I haven't got the chance to use it but I'll let you know if I have any issues. From the looks of it it's 1D position only? For my use, I have a 3D coordinate (xyz). If you feel like this can be helpful but not immediately of interest, I can contribute something later when I have more time.

**Comment by lucidrains** [comment]
@Aceticia ohh, didn't know you were working with 3d data

is alibi even suitable for this given it is symmetric? are you using Manhattan distance?

**Comment by Aceticia** [comment]
I'm not sure so I'm trying this out - my data is actually 4D, so it's actually `xyzt`. I find alibi on `t` to be pretty helpful for generalization despite it being symmetric; but I haven't figured out a good way to do `xyz` yet.

**Comment by lucidrains** [comment]
@Aceticia yes i see, alibi on time is acceptable

i think you should do [axial rotary positions](https://github.com/lucidrains/rotary-embedding-torch?tab=readme-ov-file#axial-rotary-embeddings)

**Comment by lucidrains** [comment]
@Aceticia otherwise, good luck with your project!

**Comment by Aceticia** [comment]
@lucidrains Thanks for the pointer. I think axial rotary embedding is great, but since I do need arbitrary position indicator I still need to tweak the code a little bit. Do you mind if I go through what I think here just in case something is wrong? If not I can move it over to your rope repo (wow they are anagrams).

I have an input of size [B, N, C] and a coordinate tensor of shape [B, N, 4]. I think what I need to do is modify the `get_axial_freqs` function [here](https://github.com/lucidrains/rotary-embedding-torch/blob/4da3d07cb3685dbb5ec7cd1cfe53b9693bc735ec/rotary_embedding_torch/rotary_embedding_torch.py#L265) to get one `freq` per coordinate dimension, which will give me [B, N, 4, C] - then I'm not quite sure how to proceed from there. Is it the case that one needs a continuous spatial dimension, rather than the mixed spatial dimension like I have here, for axial rope like this to work?

I guess if this is indeed the case, splitting alibi heads per axis might be the best alternative here?

**Comment by Aceticia** [comment]
@lucidrains Sorry I've been rambling quite a lot here. I decided to just try out using LieRE since it's very easy to extend to high-d cases. I adapted your code and the custom position version [is in this gist](https://gist.github.com/Aceticia/7c9b2b7df8766d8c38b24b1baefbaec0).

**Comment by lucidrains** [comment]
@Aceticia nice! how did you find that 🤣 do let me know how well it works for 3d

**Comment by lucidrains** [comment]
splitting dimensions across axis is something you could write a small paper about. it is not well studied

I think we will have to, as images and videos become small data and we reach for 4-5d

**Comment by lucidrains** [comment]
or in the far future, attention head dimensions easily hit 256, and that's enough to cram axial rotary embeddings

**Comment by Aceticia** [comment]
<img width="294" alt="Screenshot 2024-11-07 at 9 56 48 PM" src="https://github.com/user-attachments/assets/5438294b-9906-4221-8453-8e85309ec211">

It looks like LieRE doesn't show any big difference compared to simply concating rff features - I guess another point to your own observation about LieRE? I'll look into splitting heads instead.

**Comment by lucidrains** [comment]
@Aceticia hey Chris, i only compared LiERE to rotary in language and image setting, but they both lost out to rotary embeddings (axial or not) within 5k steps

**Comment by lucidrains** [comment]
@Aceticia yes, def try splitting heads, i've only heard of a few independent researchers trying this with mixed results

**Comment by lucidrains** [comment]
@Aceticia i'll also put this out there, but if you wish to forgo flash attention, use dynamic positional bias. it will definitely work, just not very efficient (but again, premature optimization is the root of all evil)

**Comment by lucidrains** [comment]
@Aceticia added the functionality for you [here](https://github.com/lucidrains/x-transformers/commit/144d9ba84955139347e798ab025457b2d7adc314#diff-10b0fcd6b3df77bf858789e5e9fb0b0662139dc0b1d66430223123062a503f8cR402)

let me know how your experiments turn out
